src/utils.jl

```diff
@@ -258,38 +258,88 @@ end
 
 # These getters improve inference since fieldtype(CodeInfo, :linetable)
 # and fieldtype(CodeInfo, :codelocs) are both Any
-const LineTypes = Union{LineNumberNode,Core.LineInfoNode}
+@static if VERSION ≥ v"1.12.0-DEV.173"
+    const LineTypes = Union{LineNumberNode,Base.IRShow.LineInfoNode}
+else
+    const LineTypes = Union{LineNumberNode,Core.LineInfoNode}
+end
 function linetable(arg)
     if isa(arg, Frame)
         arg = arg.framecode
     end
     if isa(arg, FrameCode)
         arg = arg.src
     end
-    return (arg::CodeInfo).linetable::Union{Vector{Core.LineInfoNode},Vector{Any}}  # issue #264
+    ci = arg::CodeInfo
+    @static if VERSION ≥ v"1.12.0-DEV.173"
+    return ci.debuginfo
+    else # VERSION < v"1.12.0-DEV.173"
+    return ci.linetable::Union{Vector{Core.LineInfoNode},Vector{Any}} # issue #264
+    end # @static if
 end
-_linetable(list::Vector, i::Integer) = list[i]::Union{Expr,LineTypes}
 function linetable(arg, i::Integer; macro_caller::Bool=false)::Union{Expr,LineTypes}
     lt = linetable(arg)
-    lineinfo = _linetable(lt, i)
+    @static if VERSION ≥ v"1.12.0-DEV.173"
+    # TODO: decode the linetable at this frame efficiently by reimplementing this here
+    # TODO: get the contextual name from the parent, rather than returning "n/a" (which breaks Cthulhu)
+    return Base.IRShow.buildLineInfoNode(lt, :var"n/a", i)[1] # ignore all inlining / macro expansion / etc :(
+    else # VERSION < v"1.12.0-DEV.173"
+    lin = lt[i]::Union{Expr,LineTypes}
     if macro_caller
-        while lineinfo isa Core.LineInfoNode && lineinfo.method === Symbol("macro expansion") && lineinfo.inlined_at != 0
-            lineinfo = _linetable(lt, lineinfo.inlined_at)
+        while lin isa Core.LineInfoNode && lin.method === Symbol("macro expansion") && lin.inlined_at != 0
+            lin = lt[lin.inlined_at]::Union{Expr,LineTypes}
         end
     end
-    return lineinfo
+    return lin
+    end # @static if
+end
+
+@static if VERSION ≥ v"1.12.0-DEV.173"
+
+function getlastline(arg)
+    debuginfo = linetable(arg)
+    while true
+        ltnext = debuginfo.linetable
+        ltnext === nothing && break
+        debuginfo = ltnext
+    end
+    lastline = 0
+    for k = 0:typemax(Int)
+        codeloc = Core.Compiler.getdebugidx(debuginfo, k)
+        line::Int = codeloc[1]
+        line < 0 && break
+        lastline = max(lastline, line)
+    end
+    return lastline
+end
+function codelocs(arg, i::Integer)
```

Comment:
For `statementnumbers` to function in v1.12, there's a need for a tool that can rebuild the full `codelocs` object from `Core.DebugInfo`, similar to what we could do just by looking at `src.codelocs` previously. Or, we may have to overhaul the `statementnumbers` algorithm.